leveldb: Win32WritableFile without memory mapping

laanwj opened the pull request:
Use a simple Win32WritableFile, equivalent to PosixWritableFile in posix_env.cc

The goal is to solve leveldb corruption issues on windows - this issue has been reported many times, see #5865, #6727, #5610, ...

In my testing with this change I was unable to cause the database to be corrupted when crashing a windows laptop in various ways using [notmyfault](https://technet.microsoft.com/en-us/sysinternals/bb963901.aspx). Without the change it happened the first time I tried.

laanwj committed:
leveldb: Win32WritableFile without memory mapping

Use a simple Win32WritableFile, equivalent to PosixWritableFile in
posix_env.cc

sipa commented:
Please submit to https://github.com/bitcoin/leveldb first?

laanwj commented:
Thought about that, but I think it gets more exposure and testing here.

sipa commented:
@laanwj Sure!

jgarzik commented:
ut ACK

dcousens commented:
concept ACK, utACK

sipa commented:
We should probably benchmark a reindex with this?

gmaxwell commented:
@sipa Yes, and if its much slower it's reason to figure out whats strictly required for the syncs. And if it's not, we could call it done.

laanwj commented:
This is the same as how it works on POSIX already, directly converted to WIN32 API. By all means, benchmark, but IMO we should apply this nevertheless (if it has no new bugs). We should first get rid of stability issues and only then about optimization. This will likely save many people from having to do a reindex *at all*.

MarcoFalke commented:
Concept ACK

laanwj committed:
leveldb: Don't do anything in Flush()

Flush() is supposed to clear application-side buffers (like fflush). As
this writable file implementation uses the OS function directly, there
are no buffers to flush.

laanwj commented:
Last commit is optional and potentially risky. From what I understand,  leveldb's WritableFiles are [supposed to do their own caching](https://github.com/bitcoin/bitcoin/blob/master/src/leveldb/include/leveldb/env.h#L209), and `flush()` is supposed to flush this cache to the OS (its implementation if `fflush(FILE*)` in POSIX). As this implementation uses OS calls directly, there is no need for that. With that reasoning, `Flush()` can be empty.

... it's of course also possible to use libc `FILE*` `fwrite` `fflush` on Windows instead of using the WIN32 API directly, this will provide local buffering, but I was afraid that extra level of buffering may introduce syncing issues.

Then again these are all performance concerns. As said above it may be wrong to worry about that here, I'm mostly concerned with this being correct.

Diapolo commented:
Isn't LevelDB dead at all? Their Github repo seemed to be not actively developed on the last time I checked.

laanwj commented:
@Diapolo That is a good observation. That's why I'm doing an attempt at maintaining it.
Can you help testing?

Diapolo commented:
@laanwj Yes, will first do a -reindex on testnet to be able to compare before and after... after that I'm going to kill the testnet instance and see if it's corrupting.

laanwj commented:
Thanks!

laanwj commented:
Here is an alternative implementation of `Win32WritableFile` that uses application-buffered libc primitives (fopen, fwrite, ...): https://github.com/laanwj/bitcoin/tree/2015_10_leveldb_win_nomap2   . Barely tested, and I'd not necessarily suggest using it instead of this, but if someone is going to benchmark it should be included.

laanwj commented:
@jonasschnelli it would be useful to have executables here, mind pointing your nightly builder at this?

jonasschnelli commented:
Here we go: https://bitcoin.jonasschnelli.ch/pulls/6917/

NicolasDorier commented:
I am so glad about it you can't imagine the time I lost because of this bug. This is the only reason I don't have a full node on my laptop. Trying it right now.

sipa commented on src/leveldb/util/env_win.cc:
```diff
@@ -442,202 +423,63 @@ void Win32RandomAccessFile::_CleanUp()
-    return ((x + y - 1) / y) * y;
+    std::wstring path;
+    ToWidePath(fname, path);
+    DWORD Flag = PathFileExistsW(path.c_str()) ? OPEN_EXISTING : CREATE_ALWAYS;
```
Any reason not to use OPEN_ALWAYS (which should open if exists, and create if not, according to https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858%28v=vs.85%29.aspx)?

sipa replied:
Oh, this code is just moved.

laanwj replied:
The goal is to have `f=fopen(filename, "wb")` semantics. It's possible that this could be improved, but indeed I just copied the code from the previous WritableFile implementation.

sipa commented on src/leveldb/util/env_win.cc:
```diff
@@ -103,39 +103,20 @@ class Win32RandomAccessFile : public RandomAccessFile
-    bool _MapNewRegion();
-    DISALLOW_COPY_AND_ASSIGN(Win32MapFile);
-    BOOL _Init(LPCWSTR Path);
+    std::string filename_;
```
Is this field used at all?

laanwj replied:
Good catch. Looking at env_posix it's supposed to be passed as 'context' argument to IOErrors, but that's not done at the moment.

sipa commented:
@laanwj We don't do small writes anyway, so I doubt an application-level cache would help.

sipa commented:
Concept ACK in any case.

laanwj commented:
> @laanwj We don't do small writes anyway, so I doubt an application-level cache would help.

We don't, but maybe leveldb does internally?

jonasschnelli commented:
I just have started a fresh node (mainnet / this PR) on a win8.1pro with enabled mcafee antivirus (SSD / 2GB RAM / 1.4 GHZ Core i5) ... will report soon.

jonasschnelli commented:
Have synced up to 318301,.. twice restarted bitcoin-qt (just to see if stop / start / checkblocks works). No problems with the database so far.

My node is still syncing (very slow, serval blocks per minutes).
But a strange problem appeared.
Qt console window says I'm no block 318301, ... looking in the debug.log i can see that im 318613...
Peers windows is updating... but somehow with missing data.

![bildschirmfoto-2015-11-03-um-10 19 21](https://cloud.githubusercontent.com/assets/178464/10904597/c4e984cc-8214-11e5-9fb7-425c4b407fa1.png)

Peers:
<img width="454" alt="bildschirmfoto 2015-11-03 um 10 18 41" src="https://cloud.githubusercontent.com/assets/178464/10904639/1ed09480-8215-11e5-89fa-f6c49fb2559d.png">

Taskmanager:
<img width="423" alt="bildschirmfoto 2015-11-03 um 10 23 31" src="https://cloud.githubusercontent.com/assets/178464/10904641/249451ae-8215-11e5-817c-b92b38ce870f.png">

laanwj commented:
@jonasschnelli that's a known issue and unrelated to these changes, see #5664

NicolasDorier commented:
Pro tip for corruption problem : Attach a disk, index the blockchain data on it, unplug savagely.

I had corruption everytimes my cat played with the ethernet cable between my computer and my network disk.

I did not managed to reproduce the problem so far by just killing the process.

sipa commented:
I'm pretty surprised that having the blockchain database on a network disk
works at all...

NicolasDorier commented:
Mapped Drive, then just pointing to SMB share. I'm not at my office, so I'll try with external drive unplugged savagely instead of real network disk as soon as I can.

sipa commented:
Well it's good to know that it seems to work *even* with a network
filesystem. But network filesystems usually offer much weaker
synchronization guarantees, so often are inappropriate for databases with
strong consistency requirements.

NicolasDorier commented:
Yes I agree, this is why I switched to a normal VM later on Microsoft Azure. But since I've moved to VM they are sometimes rebooted automatically for maintainance purposes, and it got corrupted several time.

The good thing about mapped disk is that thanks to weaker synchronization guarantee, it makes easier to reproduce data corruption problems ! :)

laanwj commented:
The issue that this change tries to improve is extremely easy to reproduce already: it happens nearly every time bitcoind or bitcoin-qt running on Windows crashes, or the Windows crashes, or the power is pulled. This happens for local filesystems on drives connected through SATA.

The PR does not purport to fix issues with external harddisks, network filesystems, corrupting cables, and other additional sources of complications. Being robust to those would be great, if realistically possible, but is not the immediate goal here.

jonasschnelli commented:
I have powered down (unclean sudden shutdown) my Win8.1 VM and restarted Windows and Bitcoin-Qt. Bitcoin-Qt started without issues and is ~ on the same height (353900).
I'll wait now a couple of minutes and will power down (unplug) the VM host system (MacMini) and see, what happens then.

Would it be worth to do the same procedure with the current master WITHOUT this PR to compare it?

laanwj commented:
> Would it be worth to do the same procedure with the current master WITHOUT this PR to compare it?

Sure. If you have a VM environment where you can take a snapshot, that's helpful, you can return to that after messing up the db.

jonasschnelli commented:
I made a snapshot before the first sudden shutdown. But just powered down the host by unplugging the power cable... Best chain could be activated and the node continues syncing. Will do some more shutdowns and see if nothing happens,... then switch to current master and try again.

jonasschnelli commented:
Now, after a "power off" (sudden shutdown), the db is corrupted (still using this PR).

<img width="1264" alt="screen shot 2015-11-04 at 14 16 41" src="https://cloud.githubusercontent.com/assets/178464/10939040/00a57576-82ff-11e5-8074-446c683073a9.png">

"Rebuild the block databases" ends with "error opening database":
<img width="1264" alt="bildschirmfoto 2015-11-04 um 14 20 15" src="https://cloud.githubusercontent.com/assets/178464/10939076/3b8d9042-82ff-11e5-9527-2676be078de2.png">

Log:
<img width="1264" alt="bildschirmfoto 2015-11-04 um 14 21 34" src="https://cloud.githubusercontent.com/assets/178464/10939105/5de1760e-82ff-11e5-99ec-cebd1ad75127.png">

laanwj commented:
Re: "bad undo data", see #6923 - it's a potential issue, but not something that could be solved by this.

jonasschnelli commented:
I now did restore serval times back to my snapshot and did some sudden power downs. ~50% it will end up with a corrupt database (100% with "bad undo data").

Will now try to run my snapshot with current master without this PR.

laanwj commented:
That's strange - I only managed to get the undo data error once in all my tries, whereas (without this PR) it produced consistent leveldb errors.

gmaxwell commented:
Difference in VM flushing behavior, I guess. We appear to be missing a FileCommit in UndoWriteToDisk-- I think we need to have synced the blocks and undo before calling the insert.

It would probably be better for performance if it wrote the block then undo, then did the syncs on both however.

laanwj committed:
leveldb: report filename in errors in Win32WritableFile

jonasschnelli commented:
With the current master (*without* this PR), after every sudden shut down, i get "Corruptions: error in middle of record".

<img width="1264" alt="bildschirmfoto 2015-11-04 um 16 48 15" src="https://cloud.githubusercontent.com/assets/178464/10943254/f537df84-8313-11e5-9b9f-0dc538df1724.png">

This PR looks after an improvement but getting #6923 for roughly 50% of power-off shutdowns.

laanwj commented:
If it improves things, that is good. 

Filed this for the leveldb repository: https://github.com/bitcoin/leveldb/pull/9

laanwj commented:
Closing the pull here, as it has been moved to the leveldb repository. It should come back to the bitcoin repository in a subtree update.
